[CNSMR-1890] Smarter diffs in view models unit tests (#197)
Babylonpartners/ios-playbook: Cookbook/Proposals/SmarterDiffs.md

# Improving view models' unit tests

* Author(s): Ilya Puchka
* Review Manager: TBD

## Introduction

> Our current approach to testing view models is described in [this doc](Cookbook/Technical-Documents/UnitTestingViewModels.md). This approach enforces the tests to follow "arrange-act-assert" pattern. If we forget about ergonomics of closures it works pretty well. But when it comes to assert section we have few options in how it can be implemented each with their own issues.

Reviewer: "this doc" -> 404

}
```

This approach works fine and makes the test expectations very explicit. But it becomes overly verbose when we have to assert a lot of states one by one. It may be tempting then to make shortcuts like in the example and check for particular elements in the states array and not for all of them which makes such tests incomplete as they will not catch issues when some other state unexpectedly changed or the number of states changed, unless we explicitly test for it. Another possible issue is that if we use index subscript on state array the test run can crash if index goes out of bounds instead of failing assertion and continue running other tests.
At the same time if a single state assert fails it's easier to understand what failed as we are comparing less values at a time.

#### Option 2 - assert all states at once

Example:

```swift
expect(states).to(equal([
    makeState(status: .idle),
    makeState(status: .loading),
    makeState(
        shouldShowNHSBanner: true,
        shouldShowHealthcheckBanner: true,
        shouldShowTestAndKitsBanner: true,
        shouldShowGPAtHandStatusCard: true,
        status: .loaded
    )
]))
```

> This eliminates the issue of the first approach when some state changes may be ignored by the test. It comes with a cost of actually creating all the intermediate states that might not be directly related to what test is testing which adds some amount of boilerplate. Another issue is that when this single assert fails it's very hard to extract what actually cuased this failure from the error message until we go and check each states separately. First option also suffers from this issue when each state consists of a lot of properties but it becomes much worse with a single assert.

Reviewer 2: ```suggestion
This eliminates the issue of the first approach when some state changes may be ignored by the test. It comes with a cost of actually creating all the intermediate states that might not be directly related to what test is testing which adds some amount of boilerplate. Another issue is that when this single assert fails it's very hard to extract what actually caused this failure from the error message until we go and check each states separately. First option also suffers from this issue when each state consists of a lot of properties but it becomes much worse with a single assert.
```

## Motivation

As demonstrated our current approaches suffer from few issues which result in not ideal developer experience when it comes to debugging test failures.

## Proposed solution

To improve the experience with debugging test failures we propose to use a smarter diffing approach as in [this library](https://github.com/krzysztofzablocki/Difference) by Krzysztof Zabłocki. The idea is that when comparing two instances we will be using `Mirror` type and `dump` function to first get stored properties and then get their textual representation and this way combine a in-memory snapshot of the value. This way we can recursively compare each property and only include those that differ in the final failure message. 

> The fact that `Mirror` only provides access to stored properties is not an issue as computed properties are products of stored properties (or in case of enums of the enum cases and thier associated values) and for same stored properties values should return the same result.

Reviewer: ```suggestion
The fact that `Mirror` only provides access to stored properties is not an issue as computed properties are products of stored properties (or in case of enums of the enum cases and their associated values) and for same stored properties values should return the same result.
```

Example:

```swift
// equalDiff is a name suggested in the Diffable repo README, we can choose a different one
expect(states).to(equalDiff([
    makeState(status: .idle),
    makeState(status: .loading),
    makeState(
        shouldShowNHSBanner: false, // this value causes failure
        shouldShowHealthcheckBanner: true,
        shouldShowTestAndKitsBanner: true,
        shouldShowGPAtHandStatusCard: true,
        status: .loaded
    )
]))

> ```

Reviewer 3: I am not convinced that writing all this boiler plate is worth it. If my test is checking a particular property in the state after an action of steps, I would like the test to be focussed on those subset of properties. It makes the reading and understanding the purpose of the test easier.
As to the problem of crashes when states array's index is out of bounds, maybe we can use guard statement
e.g. 
```
guard states.count > 10 else { XCTFail()} 
expect(states.first?.lastName).to(beNil())
```

Author: There are benefits and tradeoffs in each of these options but regardless which one we use they both benefit from the proposal

Reviewer 4: @chitrakotwani But if you only assert on the subset of properties in question, then you will miss catching the corruption to state of properties outside the subset. Its best to assert the entire state equals your expected state. You could differentiate the setup of the expectation however with only the subset of properties (to understand what you are concerned about in the specific test case).

Reviewer 3: @anilputtabuddhi - This would make the tests more brittle because an unexpected change in one state property would break quite a few tests at once. This would make it difficult to pinpoint the blame of breaking tests.

But as @ilyapuchka  suggests, this is probably out of scope of discussion of this proposal anyway.

Author: Agree with @anilputtabuddhi. We will always compromise between coverage and clarity of tests. I believe we should prefer coverage but have a clarity in mind (test should be steal readable and understandable). That's not what the proposal is about though =)

Reviewer 4: @chitrakotwani The tests would not be called brittle in the scenario you explained - An unexpected state change should cause failures and thats what tests are for. A brittle test is one that fails and succeeds intermittently on different runs. It is also possible to architect tests with proper fixture designs that help in test setup to be able to pinpoint breakages as you say.

Reviewer 3: @anilputtabuddhi - Brittle tests, as name implies, are the ones which break easily. If a test will begin to depend on too many state's properties, they will start breaking easily. A small change somewhere will have unintended effects in tests in other areas. 
Tests which fail and succeed intermittently on different runs could be called unreliable?

```

> As can be seen from the diff it does not show in what exact array item the faulure happend. This is something that we will need to adjust along with [this issue](https://github.com/krzysztofzablocki/Difference/issues/1). 

Reviewer 5: ```suggestion
As can be seen from the diff it does not show in what exact array item the failure happened. This is something that we will need to adjust along with [this issue](https://github.com/krzysztofzablocki/Difference/issues/1). 
```

> Proposed way to integrate this feature is to copy the implementation (with proper author attribution) and adjust it to our needs, optionally we can make pull requests to the original library.

Reviewer 6: Having a local copy might make it a bit easier to work on fixing the issues with library. Once we're confident in our fixes, we could submit PR to original library. Not sure how much of an overhead it would be to maintain our fork.

## Impact on existing codebase

> This approach can be rolled out gradually and used for new test (Xcode templates will use a new assert) or tests can be refactored in batches by replacing existing assertions with a new assert.

Reviewer 6: Given all issues with library, see my main review comment, I'd make it a non-default assertion. Something we can use when tests we're working on fails.
```swift
/// keep using `equal` for state tests
expect(states).to(equal([
    makeState(status: .idle),
    ...
)
// and if above fails we can switch to reflection based message for debugging or development
expect(states).to(equalWithDiff([
    makeState(status: .idle),
    ...
)
```

I'd keep it this way until we have more experience with it and have good idea in what areas we'd need to improve messages.

Author: Yes, probably.
The issues you highlighted seem all to relate to arrays to items, so we can try to mitigate them by applying diff algorithm after we apply reflection so that it will calculate mutations - if there are no mutations needed - assert passes, otherwise we see how things moved around.

## Alternatives considered

> As an alternative we can enforce asserting on individual states instead of all states which will make diffs less of an issue. But as mentioned before it will not eliminate it.

Reviewer 6: I like the idea of seeing all the states. It makes it a little bit easier to figure out how given view model works. Plus, those tests are already a bit broad, so I'd say it's fine. Definitely agree with downsides you pointed out in "Option 1 - assert states one by one".

> As an alternative we can inforce asserting on individual states insteead of all states which will make diffs less of an issue. But as mentioned before it will not eliminate it.

Reviewer 2: ```suggestion
As an alternative we can enforce asserting on individual states instead of all states which will make diffs less of an issue. But as mentioned before it will not eliminate it.
```

> Another alternative would be to use snapshot testing that we already use for UI tests. It will make our assert just a call to a single `assertSnapshot` function. But at the same time it will mean that all the snapshots will be written into the files rather than explicitly stated in the test code. This will decrease the boilerplate we will have to write in tests manually but will make it harder to understand test expectations as as we will need to search through text files to find values we need.

Reviewer 6: I wonder if it would be possible to use snapshot library infrastructure to capture diffs on the fly in memory without storing them in files. It would work like `Difference` as in produce snapshots only on failure.
Would require some more usage to see how useful are the diffs it provides. I think we have less experience with this than with `Difference`, so it's hard to tell.

Author: The convenience of snapshots is that we don't have to specify the expected values since they are in written the file. If we don't write them then we have to construct them on every run so we end up with the same as regular asserts, maybe a bit different API.

> Another alternative would be to use snapshot testing that we already use for UI tests. It will make our assert just a call to a single `assertSnapshot` function. But at the same time it will mean that all the snapshots will be writtent into the files rather than explicitly stated in the test code. This will decrease the boilerplate we will have to write in tests manually but will make it harder to understand test expectations as as we will need to search through text files to find values we need.

Reviewer 2: ```suggestion
Another alternative would be to use snapshot testing that we already use for UI tests. It will make our assert just a call to a single `assertSnapshot` function. But at the same time it will mean that all the snapshots will be written into the files rather than explicitly stated in the test code. This will decrease the boilerplate we will have to write in tests manually but will make it harder to understand test expectations as as we will need to search through text files to find values we need.
```

Reviewer: Taking @brocoo 's suggestion + also removing a duplicate `as`
```suggestion
Another alternative would be to use snapshot testing that we already use for UI tests. It will make our assert just a call to a single `assertSnapshot` function. But at the same time it will mean that all the snapshots will be written into the files rather than explicitly stated in the test code. This will decrease the boilerplate we will have to write in tests manually but will make it harder to understand test expectations as we will need to search through text files to find values we need.
```
